text/0000-normalized-modules.md

```diff
@@ -0,0 +1,535 @@
+
+One more rule is proposed to simplify file naming: top-level modules
+named `main` can omit their name and just be named by `type` without
+ambiguity:
+
+* `modules/${type}` - `name` is assumed to be `main` when omitted
+
+Instead of `modules/main/app.js` or `modules/main.app.js`, the
+simpler alternative `modules/app.js` can be allowed.
+
+#### Organize files in "buckets"
+
+One last optimization is to reduce the noise caused by directories filled with
+files of all different types.
+
+It's proposed that we introduce special "bucket" directories for organizing
+files. These directories should be prefixed with a `-` to indicate that they
+will not be used to inform module names, namespaces, or types. Every module
+defined within a bucket simply flows through to the bucket's parent namespace.
+
+Ember CLI should probably enforce the use of these buckets as well as the types
+of modules allowed in each bucket.
+
+The following buckets are proposed:
+
+* `-components` - components and their corresponding templates
+* `-data` - including models, adapters, and serializers
+* `-helpers` - helpers
+* `-initializers` - application and instance initializers
+* `-routes` - routes, templates, and routable components
+* `-services` - services
+
+### Files for an example application
+
+Given the base and alternative _file_ naming patterns, the example application
+could be restructured using the following files and directories:
+
+```
+modules
+  -routes
+    posts
+      new
+        -components
+          new-post
+            component.js
+            template.hbs
+        route.js
+        template.hbs
+      -components
+        post-editor
+          -helpers
+            calculate-post-title.helper.js
+          post-editor-button
+            component.js
+            template.hbs
+          component.js
+          template.hbs
+        post-viewer
+          component.js
+          template.hbs
+      -helpers
+        capitalize.helper.js
+        titleize.helper.js
+      route.js
+      template.hbs
+  -components
+    list-paginator
+      component.js
+      template.hbs
+  -data
+    comments
+      adapter.js
+      model.js
+      serializer.js
+    posts
+      adapter.js
+      model.js
+      serializer.js
+  -helpers
+    capitalize.helper.js
+  -initializers
+    auth.instance-initializer.js
+    i18n.initializer.js
+  -services
+    auth.service.js
+  app.js
+  router.js
+  template.hbs
+```
+
+This example shows that we can improve the ergonomics of naming _files_ in
+Ember CLI projects, while normalizing everything into a single _module_
+naming structure.
+
+## Normalized module:file map
+
+The Ember CLI resolver will need to maintain a map of normalized module
+names to file locations on disk. This map will provide a single place for
+module lookups by the resolver.
+
+Construction of this map will allow for detection of any module name conflicts.
+In the case of any conflicts, Ember CLI should throw an exception.
```

Comment:
Yep, we absolutely plan to do this at build time so the runtime resolver has much much less work to do.